File: myulib-core/src/main/java/com/myudog/myulib/internal/generator/CylinderGenerator.java
package com.myudog.myulib.internal.generator;
import com.myudog.myulib.api.Shapes;
import com.myudog.myulib.api.core.shape.IShape;
import net.minecraft.world.phys.Vec3;import java.util.ArrayList;
import java.util.List;
public final class CylinderGenerator implements IShape {
    @Override
    public List<Vec3> getOutlinePoints(Vec3 size, double density) {
        List<Vec3> points = new ArrayList<>();
        double radiusX = size.x;
        double radiusZ = size.z;
        double height = size.y;
        int circlePoints = Math.max(8, (int) Math.round((radiusX + radiusZ) * Math.PI * Math.max(1.0, density)));
        for (int i = 0; i < circlePoints; i++) {
            double angle = 2.0 * Math.PI * i / circlePoints;
            double x = radiusX * Math.cos(angle);
            double z = radiusZ * Math.sin(angle);
            points.add(new Vec3(x, 0.0, z));
            points.add(new Vec3(x, height, z));
        }
        int verticalSteps = Math.max(1, (int) Math.round(height * Math.max(1.0, density)));
        for (int i = 0; i < verticalSteps; i++) {
            double h = i / Math.max(1.0, density);
            for (int j = 0; j < 4; j++) {
                double angle = j * Math.PI / 2.0;
                points.add(new Vec3(radiusX * Math.cos(angle), h, radiusZ * Math.sin(angle)));
            }
        }
        return points;
    }
    @Override
    public List<Vec3> getSolidPoints(Vec3 size, double density) {
        List<Vec3> points = new ArrayList<>();
        double hStep = 1.0 / Math.max(0.1, density);
        for (double h = 0.0; h <= size.y; h += hStep) {
            for (Vec3 point : Shapes.CIRCLE.getSolidPoints(new Vec3(size.x, 0.0, size.z), density)) {
                points.add(new Vec3(point.x, h, point.z));
            }
        }
        return points;
    }
}

File: src/main/kotlin/com/myudog/myulib/internal/generator/CylinderGenerator.kt
package com.myudog.myulib.internal.generator

import com.myudog.myulib.api.shape.IShape
import net.minecraft.util.math.Vec3d
import kotlin.math.*

internal class CylinderGenerator : IShape {
    override fun getOutlinePoints(size: Vec3d, density: Double): List<Vec3d> {
        val points = mutableListOf<Vec3d>()
        val radiusX = size.x
        val radiusZ = size.z
        val height = size.y

        // 1. 生成上下兩個圓盤的邊框
        val circlePoints = (radiusX + radiusZ) * PI * density // 粗略估算點數
        val step = 2 * PI / circlePoints

        for (i in 0 until circlePoints.toInt()) {
            val angle = i * step
            val x = radiusX * cos(angle)
            val z = radiusZ * sin(angle)
            points.add(Vec3d(x, 0.0, z)) // 底部
            points.add(Vec3d(x, height, z)) // 頂部
        }

        // 2. 生成側面的垂直線線 (以密度為準)
        for (i in 0 until (height * density).toInt()) {
            val h = i / density
            // 在側面均勻選四個點連線 (或更多)
            for (j in 0..3) {
                val angle = j * PI / 2
                points.add(Vec3d(radiusX * cos(angle), h, radiusZ * sin(angle)))
            }
        }
        return points
    }

    override fun getSolidPoints(size: Vec3d, density: Double): List<Vec3d> {
        val points = mutableListOf<Vec3d>()
        val height = size.y
        // 均勻填充圓柱 (分層填充)
        var h = 0.0
        val hStep = 1.0 / density
        while (h <= height) {
            // 每一層使用 CircleGenerator 的 getSolidPoints
            points.addAll(com.myudog.myulib.api.Shapes.CIRCLE.getSolidPoints(Vec3d(size.x, 0.0, size.z), density))
            points.replaceAll { if (it.y == 0.0) Vec3d(it.x, h, it.z) else it } // 修正高度
            h += hStep
        }
        return points
    }
}